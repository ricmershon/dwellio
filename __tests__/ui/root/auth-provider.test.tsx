import React from 'react';
import { render, screen } from '@/__tests__/test-utils';

import AuthProvider from '@/ui/root/auth-provider';

// Mock NextAuth SessionProvider
jest.mock('next-auth/react', () => ({
    SessionProvider: ({ children }: { children: React.ReactNode }) => (
        <div data-testid="session-provider">
            {children}
        </div>
    ),
}));

describe('AuthProvider', () => {
    describe('Component Structure', () => {
        it('should render children within SessionProvider', () => {
            const TestChild = () => <div data-testid="test-child">Test Child Component</div>;
            
            render(
                <AuthProvider>
                    <TestChild />
                </AuthProvider>
            );
            
            expect(screen.getByTestId('session-provider')).toBeInTheDocument();
            expect(screen.getByTestId('test-child')).toBeInTheDocument();
        });

        it('should wrap children with NextAuth SessionProvider', () => {
            const TestChild = () => <div data-testid="wrapped-child">Child</div>;
            
            render(
                <AuthProvider>
                    <TestChild />
                </AuthProvider>
            );
            
            const sessionProvider = screen.getByTestId('session-provider');
            const childElement = screen.getByTestId('wrapped-child');
            
            expect(sessionProvider).toBeInTheDocument();
            expect(childElement).toBeInTheDocument();
            expect(sessionProvider).toContainElement(childElement);
        });
    });

    describe('Children Rendering', () => {
        it('should render single child component', () => {
            render(
                <AuthProvider>
                    <div data-testid="single-child">Single Child</div>
                </AuthProvider>
            );
            
            expect(screen.getByTestId('single-child')).toBeInTheDocument();
            expect(screen.getByText('Single Child')).toBeInTheDocument();
        });

        it('should render multiple child components', () => {
            render(
                <AuthProvider>
                    <div data-testid="first-child">First Child</div>
                    <div data-testid="second-child">Second Child</div>
                    <span data-testid="third-child">Third Child</span>
                </AuthProvider>
            );
            
            expect(screen.getByTestId('first-child')).toBeInTheDocument();
            expect(screen.getByTestId('second-child')).toBeInTheDocument();
            expect(screen.getByTestId('third-child')).toBeInTheDocument();
        });

        it('should render complex nested children', () => {
            render(
                <AuthProvider>
                    <div data-testid="parent">
                        <div data-testid="nested-child">
                            <span data-testid="deeply-nested">Deep Content</span>
                        </div>
                    </div>
                </AuthProvider>
            );
            
            expect(screen.getByTestId('parent')).toBeInTheDocument();
            expect(screen.getByTestId('nested-child')).toBeInTheDocument();
            expect(screen.getByTestId('deeply-nested')).toBeInTheDocument();
            expect(screen.getByText('Deep Content')).toBeInTheDocument();
        });

        it('should handle empty children gracefully', () => {
            const { container } = render(
                <AuthProvider>
                    {null}
                </AuthProvider>
            );
            
            expect(screen.getByTestId('session-provider')).toBeInTheDocument();
            expect(container.firstChild).toBeInTheDocument();
        });

        it('should render text content', () => {
            render(
                <AuthProvider>
                    Plain text content
                </AuthProvider>
            );
            
            expect(screen.getByText('Plain text content')).toBeInTheDocument();
        });

        it('should render mixed content types', () => {
            render(
                <AuthProvider>
                    <div data-testid="element">Element</div>
                    Text content
                    <span data-testid="another-element">Another Element</span>
                </AuthProvider>
            );
            
            expect(screen.getByTestId('element')).toBeInTheDocument();
            expect(screen.getByText('Text content')).toBeInTheDocument();
            expect(screen.getByTestId('another-element')).toBeInTheDocument();
        });
    });

    describe('Props Handling', () => {
        it('should accept children as prop with correct typing', () => {
            const children = <div data-testid="typed-child">Typed Child</div>;
            
            render(<AuthProvider>{children}</AuthProvider>);
            
            expect(screen.getByTestId('typed-child')).toBeInTheDocument();
        });

        it('should handle readonly children prop', () => {
            // This test ensures the Readonly<{ children: React.ReactNode }> typing works
            const TestComponent = () => (
                <AuthProvider>
                    <div data-testid="readonly-test">Readonly Children Test</div>
                </AuthProvider>
            );
            
            render(<TestComponent />);
            
            expect(screen.getByTestId('readonly-test')).toBeInTheDocument();
        });
    });

    describe('SessionProvider Integration', () => {
        it('should pass children to SessionProvider correctly', () => {
            const childContent = 'Session Provider Child';
            
            render(
                <AuthProvider>
                    <div>{childContent}</div>
                </AuthProvider>
            );
            
            const sessionProvider = screen.getByTestId('session-provider');
            expect(sessionProvider).toBeInTheDocument();
            expect(screen.getByText(childContent)).toBeInTheDocument();
        });

        it('should maintain component hierarchy', () => {
            render(
                <AuthProvider>
                    <div data-testid="app-content">
                        <div data-testid="main-content">Main App</div>
                    </div>
                </AuthProvider>
            );
            
            const sessionProvider = screen.getByTestId('session-provider');
            const appContent = screen.getByTestId('app-content');
            const mainContent = screen.getByTestId('main-content');
            
            expect(sessionProvider).toBeInTheDocument();
            expect(appContent).toBeInTheDocument();
            expect(mainContent).toBeInTheDocument();
            
            // Verify hierarchy
            expect(sessionProvider).toContainElement(appContent);
            expect(appContent).toContainElement(mainContent);
        });
    });

    describe('Component Behavior', () => {
        it('should function as a wrapper component', () => {
            let childRenderCount = 0;
            
            const TestChild = () => {
                childRenderCount++;
                return <div data-testid="counting-child">Render count: {childRenderCount}</div>;
            };
            
            render(
                <AuthProvider>
                    <TestChild />
                </AuthProvider>
            );
            
            expect(screen.getByTestId('counting-child')).toBeInTheDocument();
            expect(screen.getByText('Render count: 1')).toBeInTheDocument();
        });

        it('should not interfere with child component functionality', () => {
            let clickCount = 0;
            
            const ClickableChild = () => (
                <button 
                    data-testid="clickable-child"
                    onClick={() => { clickCount++; }}
                >
                    Click me
                </button>
            );
            
            render(
                <AuthProvider>
                    <ClickableChild />
                </AuthProvider>
            );
            
            const button = screen.getByTestId('clickable-child');
            expect(button).toBeInTheDocument();
            
            // Simulate click to ensure functionality is preserved
            button.click();
            expect(clickCount).toBe(1);
        });

        it('should preserve React keys and other properties', () => {
            const items = [1, 2, 3];
            
            render(
                <AuthProvider>
                    {items.map(item => (
                        <div key={item} data-testid={`item-${item}`}>
                            Item {item}
                        </div>
                    ))}
                </AuthProvider>
            );
            
            items.forEach(item => {
                expect(screen.getByTestId(`item-${item}`)).toBeInTheDocument();
                expect(screen.getByText(`Item ${item}`)).toBeInTheDocument();
            });
        });
    });

    describe('Error Boundaries', () => {
        it('should handle children that throw errors gracefully', () => {
            // Suppress console.error for this test
            const originalError = console.error;
            console.error = jest.fn();
            
            const ThrowingChild = () => {
                throw new Error('Test error');
            };
            
            expect(() => {
                render(
                    <AuthProvider>
                        <ThrowingChild />
                    </AuthProvider>
                );
            }).toThrow('Test error');
            
            // Restore console.error
            console.error = originalError;
        });
    });

    describe('Accessibility', () => {
        it('should not add any accessibility barriers', () => {
            render(
                <AuthProvider>
                    <button aria-label="Accessible button">Click me</button>
                    <div role="main">Main content</div>
                </AuthProvider>
            );
            
            expect(screen.getByLabelText('Accessible button')).toBeInTheDocument();
            expect(screen.getByRole('main')).toBeInTheDocument();
        });

        it('should preserve ARIA attributes in children', () => {
            render(
                <AuthProvider>
                    <div 
                        role="navigation" 
                        aria-label="Main navigation"
                        data-testid="nav"
                    >
                        Navigation content
                    </div>
                </AuthProvider>
            );
            
            const nav = screen.getByTestId('nav');
            expect(nav).toHaveAttribute('role', 'navigation');
            expect(nav).toHaveAttribute('aria-label', 'Main navigation');
        });
    });

    describe('Snapshots', () => {
        it('should match snapshot with simple children', () => {
            const { container } = render(
                <AuthProvider>
                    <div>Simple child content</div>
                </AuthProvider>
            );
            
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot with complex children', () => {
            const { container } = render(
                <AuthProvider>
                    <div className="app">
                        <header>Header</header>
                        <main>
                            <section>Content</section>
                        </main>
                        <footer>Footer</footer>
                    </div>
                </AuthProvider>
            );
            
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot with no children', () => {
            const { container } = render(<AuthProvider>{null}</AuthProvider>);
            
            expect(container.firstChild).toMatchSnapshot();
        });
    });
});