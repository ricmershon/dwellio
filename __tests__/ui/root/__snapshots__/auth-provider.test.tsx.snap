// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AuthProvider Snapshots should match snapshot with complex children 1`] = `
<div
  data-testid="session-provider"
>
  <div
    class="app"
  >
    <header>
      Header
    </header>
    <main>
      <section>
        Content
      </section>
    </main>
    <footer>
      Footer
    </footer>
  </div>
</div>
`;

exports[`AuthProvider Snapshots should match snapshot with no children 1`] = `
<div
  data-testid="session-provider"
/>
`;

exports[`AuthProvider Snapshots should match snapshot with simple children 1`] = `
<div
  data-testid="session-provider"
>
  <div>
    Simple child content
  </div>
</div>
`;
